pgmpy/models/FunctionalBayesianNetwork.py

```diff
@@ -191,3 +192,156 @@ def simulate(self, n_samples=1000, seed=None):
+            num_steps (int) : Number of optimization steps for each variable.
+
+        Returns:
+            dict: Samples for each register parameter.
+
+        Examples
+        --------
+        >>> from pgmpy.factors.hybrid import FunctionalCPD
+        >>> from pgmpy.models import FunctionalBayesianNetwork
+        >>> import numpy as np
+        >>> import pyro.distributions as dist
+
+        >>> model = FunctionalBayesianNetwork([("x1", "x2")])
+        >>> x1 = np.random.normal(1, 2, size=10000)
+        >>> x2 = np.random.normal(5 + x1, 1)
+        >>> data = pd.DataFrame({"x1": x1, "x2": x2})
+        >>> def x1_prior():
+        ...    mu = pyro.sample("x1_mu", dist.Normal(0, 10))
+        ...    sigma = pyro.sample("x1_sigma", dist.HalfNormal(5))
+        ...    return dist.Normal(mu, sigma)
+        >>> def x2_prior(parent):
+        ...    mu = pyro.param("x2_mu", torch.tensor(1.0)) + parent["x1"]
+        ...    sigma = positive_param("x2_sigma", 1.0)
+        ...    return dist.Normal(mu, sigma)
+
+        >>> cpd1 = FunctionalCPD("x1", lambda _: x1_prior())
+        >>> cpd2 = FunctionalCPD('x2', fn=lambda parent: x2_prior(parent), parents=['x1'])
+        >>> model.add_cpds(cpd1, cpd2)
+        >>> params = model.fit(data, method="SVI", learning_rate=0.05, num_steps=100)
+        >>> print(params)
+
+        >>> def x1_prior():
+        ...    mu = pyro.sample("x1_mu", dist.Normal(0, 10))
+        ...    sigma = pyro.sample("x1_sigma", dist.HalfNormal(5))
+        ...    return dist.Normal(mu, sigma)
+
+        >>> def x2_prior(parent):
+        ...    mu = pyro.sample("x2_mu", dist.Normal(5, 1))
+        ...    sigma = pyro.sample("x2_sigma", dist.HalfNormal(2))
+        ...    return dist.Normal(mu + parent['x1'], sigma)
+
+        >>> cpd1 = FunctionalCPD("x1", lambda _: x1_prior())
+        >>> cpd2 = FunctionalCPD('x2', fn=lambda parent: x2_prior(parent), parents=['x1'])
+
+        >>> params = model.fit(data, method="MCMC", num_steps=100, mcmc_kwargs={"mp_context": "fork"})
+        >>> print(params["x1_mu"].mean(), params["x1_std"].mean())
+        """
+
+        if not isinstance(data, pd.DataFrame):
+            raise ValueError("Specify data as a pandas Dataframe.")
+        if not isinstance(learning_rate, float):
+            raise ValueError(
+                f"Learning rate should be float type, not {type(learning_rate)}"
+            )
+        if not isinstance(num_steps, int):
+            raise ValueError(
+                f"Number of steps should be int type, not {type(num_steps)}"
+            )
+        if method not in ["SVI", "MCMC"]:
+            raise ValueError("Current implementation only support SVI or MCMC.")
+
+        sort_nodes = list(nx.topological_sort(self))
+
+        if method == "SVI":
+
+            def model():
+                with pyro.plate(f"data", data.shape[0]):
+                    for node in self.nodes:
+                        if node in data.columns:
```

Comment:
I think we should move this check outside of the fitting loop. We should throw an error if there are missing nodes in the data at the start of the method.

Reply:
@Nuna7 No, when fitting the model we are trying to learn the parameters for all the variables. So, we would need data for all of them. There would be missing variables (as in your example) only for inference.